cmd/ecc/join.go

```diff
@@ -0,0 +1,330 @@
+// it tries to fetch any parity shards. Minimizing
+// the number of reads / connections is important to
+// avoid any unnecessary I/O or network load. Also, if
+// a data shard is available we should prefer fetching it
+// instead of reading a parity shard to avoid unnecessary
+// erasure coding reconstruction.
+func (r *JoinedReaders) Read(b *Buffer) error {
+	if len(r.src) != len(b.shards) {
+		panic("ecc: number of data sources does not match the number of shards")
+	}
+	if r.rErr != nil {
+		return r.rErr
+	}
+
+	b.Reset()
+	results, next, err := r.spawn(b)
+	if err != nil {
+		r.rErr = err
+		return r.rErr
+	}
+
+	// Mark all shards as missing.
+	// When a read from a data source succeeds
+	// the particular shard will get marked
+	// as available again.
+	for i := range b.shards {
+		b.shards[i] = b.shards[i][:0]
+	}
+
+	return r.join(b, next, results)
+}
+
+type readResult struct {
+	ID  int   // The ID / index of the shard / data source
+	N   int   // The number of bytes read from the data source
+	Err error // Any error that happened during reading
+}
+
+// spawn creates just as many concurrent go routines as necessary to read
+// enough shards to reconstruct the content data. If all data shards are
+// available each go routine tries to read a data shard.
+// It returns a channel which hold the read results once they are available.
+// The returned next index points to the next data source from which a go
+// routine should read if any of the spawned reads fails.
+func (r *JoinedReaders) spawn(b *Buffer) (results chan readResult, next int, err error) {
+	results = make(chan readResult, len(r.src))
+	var spawned int
+	for spawned < len(b.data) && next < len(b.shards) {
+		if r.err[next] == nil {
+			go readFrom(r.src[next], next, r.offset, b.shards[next], results)
+			spawned++
+		}
+		next++
+	}
+	// If we could not spwan a single go routine we should return an error
+	// (or panic) since a join would block forever since the results channel
+	// would never receive a result. However, that error should never happen
+	// since there should be at least 1 (actually `len(b.data)`) non-nil data
+	// sources. Otherwise, we never had read quorum in the first place and
+	// should have failed somewhere higher up the stack.
+	// TODO(aead): Consider a panic.
+	if spawned == 0 {
+		err = ErrReadQuorum
+	}
+	return
+}
+
+// join joins all go routines created by spawn(). Therefore, it reads
+// from the results channel until it has received enough shards. If
+// one read fails - i.e. returns an non-EOF error - it re-spawns another
+// go routine reading from the next data source.
+func (r *JoinedReaders) join(b *Buffer, next int, results chan readResult) error {
+	var (
+		n      int // Number of bytes read in total
+		nReads int // Number of reads that succeeded
+		nEOFs  int // Number of cases read returned an EOF error
+	)
+	for result := range results {
+		i, nn, err := result.ID, result.N, result.Err
```

Comment:
```suggestion
		id, nn, err := result.ID, result.N, result.Err
```
Let's keep `i` for loop counters.

Reply:
Which it is - but fair enough :wink:

Reply:
ah, ok didn't realize :)